flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategyTest.java

```diff
@@ -0,0 +1,248 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link TaskBalancedPreferredSlotSharingStrategy}. */
+class TaskBalancedPreferredSlotSharingStrategyTest {
+
+    @Test
+    void testCoLocationConstraintIsRespected() {
+        TestingSchedulingTopology topology = new TestingSchedulingTopology();
+        List<Tuple2<JobVertexID, List<TestingSchedulingExecutionVertex>>> jobVertexInfos =
+                new ArrayList<>();
+        SlotSharingGroup slotSharingGroup1 = new SlotSharingGroup();
+        CoLocationGroup coLocationGroup1 = new CoLocationGroupImpl();
+        CoLocationGroup coLocationGroup2 = new CoLocationGroupImpl();
+        List<TestingJobVertexInfo> mockedJobVertices =
+                getMockedJobVertices(slotSharingGroup1, coLocationGroup1, coLocationGroup2);
+        setupCase(mockedJobVertices, topology, jobVertexInfos);
+
+        final SlotSharingStrategy strategy =
+                new TaskBalancedPreferredSlotSharingStrategy(
+                        topology,
+                        Sets.newHashSet(slotSharingGroup1),
+                        Sets.newHashSet(coLocationGroup1, coLocationGroup2));
+        List<TestingSchedulingExecutionVertex> executionVertices1 = jobVertexInfos.get(1).f1;
+        List<TestingSchedulingExecutionVertex> executionVertices2 = jobVertexInfos.get(2).f1;
+        TestingSchedulingExecutionVertex executionVertexOfJv0 = jobVertexInfos.get(0).f1.get(0);
+        ExecutionSlotSharingGroup executionSlotSharingGroupOfExecutionVertexOfJv0 =
+                strategy.getExecutionSlotSharingGroup(executionVertexOfJv0.getId());
+
+        assertThat(executionVertices1).hasSameSizeAs(executionVertices2);
+        for (int i = 0; i < executionVertices1.size(); i++) {
+            ExecutionSlotSharingGroup executionSlotSharingGroup =
+                    strategy.getExecutionSlotSharingGroup(executionVertices1.get(i).getId());
+            assertThat(executionSlotSharingGroup)
+                    .isNotEqualTo(executionSlotSharingGroupOfExecutionVertexOfJv0);
+            assertThat(executionSlotSharingGroup)
+                    .isEqualTo(
+                            strategy.getExecutionSlotSharingGroup(
+                                    executionVertices2.get(i).getId()));
+        }
+
+        List<TestingSchedulingExecutionVertex> executionVertices3 = jobVertexInfos.get(3).f1;
+        List<TestingSchedulingExecutionVertex> executionVertices4 = jobVertexInfos.get(4).f1;
+        for (int i = 0; i < executionVertices1.size(); i++) {
+            assertThat(strategy.getExecutionSlotSharingGroup(executionVertices3.get(i).getId()))
+                    .isEqualTo(
+                            strategy.getExecutionSlotSharingGroup(
+                                    executionVertices4.get(i).getId()));
+        }
+    }
+
+    @Nonnull
+    private static ArrayList<TestingJobVertexInfo> getMockedJobVertices(
```

Comment:
If it is only used in one test, no need to split them.

Reply:
inlined it back into the calling lines.